pkg/ccl/importccl/read_import_base.go

```diff
@@ -153,24 +153,54 @@ func readInputFiles(
 
 	fileSizes := make(map[int32]int64, len(dataFiles))
 
-	// Attempt to fetch total number of bytes for all files.
+	// "Pre-import" work.
+	// Validate permissions early, and attempt to fetch total number of bytes for all files to track progress.
 	for id, dataFile := range dataFiles {
-		conf, err := cloud.ExternalStorageConfFromURI(dataFile, user)
-		if err != nil {
-			return err
-		}
-		es, err := makeExternalStorage(ctx, conf)
-		if err != nil {
+		if err := func() error {
+			conf, err := cloud.ExternalStorageConfFromURI(dataFile, user)
+			if err != nil {
+				return err
+			}
+			es, err := makeExternalStorage(ctx, conf)
+			if err != nil {
+				return err
+			}
+			defer es.Close()
+
+			sz, err := es.Size(ctx, "")
+
+			if sz <= 0 {
+				// Don't log dataFile here because it could leak auth information.
+				log.Infof(ctx, "could not fetch file size; falling back to per-file progress: %v", err)
+				return nil
```

Comment:
I'm not sure that a rejected HEAD request should imply we don't move on to check readability using a GET, since those seem separate?

Reply:
I guess following that, if we switch this `return nil` to an `else { fileSizes[id...` then I also start to wonder if the anon-function-as-value-yielding-block is needed, vs just inverting the `if` below to remove the only other early nil return.

Reply:
Fixed the failed-HEAD, thanks for catching! I'd like to keep the anon function for the sake of the defers, which otherwise don't work quite right in a for loop I think. Seem reasonable?

Reply:
I'm not sure you pushed the amended commit yet, but sure, that sounds fine.

Re defer in loop with/without anon function it still "works" just fine, it just changes when it'd run -- they'd all be queued up to run at the end of the outer function vs after each loop, but they'll still run. I'm fine with it either way though. Maybe a comment though to explain the anon function is to fun the deferred close immediately.